strawberryfields/backends/bosonicbackend/backend.py

```diff
@@ -54,6 +69,205 @@ def __init__(self):
+
+                    applied.append(cmd)
+
+                except NotApplicableError as e:
+                    # command is not applicable to the current backend type
+                    raise NotApplicableError(
+                        "The operation {} cannot be used with the Bosonic Backend.".format(cmd.op)
+                    ) from e
+
+                except NotImplementedError as e:
+                    # command not directly supported by backend API
+                    raise NotImplementedError(
+                        "The operation {} has not been implemented in the Bosonic Backend for the arguments {}.".format(
+                            cmd.op, kwargs
+                        )
+                    ) from e
+
+        return applied, samples_dict, all_samples
+
+    def init_circuit(self, prog):
+        """Instantiate the circuit and initialize weights, means, and covs
+        depending on the Preparation classes.
+
+        Args:
+            prog (object): sf.Program instance
+
+        Raises:
+            NotImplementedError: if ``Ket`` or ``DensityMatrix`` preparation used
+        """
+        from strawberryfields.ops import (
+            Bosonic,
+            Catstate,
+            DensityMatrix,
+            Fock,
+            GKP,
+            Ket,
+            _New_modes,
+        )
+
+        nongausspreps = (Bosonic, Catstate, DensityMatrix, Fock, GKP, Ket)
+        nmodes = prog.init_num_subsystems
+        self.begin_circuit(nmodes)
+        # Dummy initial weights, means and covs
+        init_weights, init_means, init_covs = [[0] * nmodes for i in range(3)]
+
+        vac_means = np.zeros((1, 2), dtype=complex)  # .tolist()
+        vac_covs = np.array([0.5 * self.circuit.hbar * np.identity(2)])
+
+        # List of modes that have been traversed through
+        reg_list = []
+
+        # Go through the operations in the circuit
+        for cmd in prog.circuit:
+            # Check if an operation other than New() has already acted on these modes.
+            labels = [label.ind for label in cmd.reg]
+            isitnew = 1 - np.isin(labels, reg_list)
+            if np.any(isitnew):
+                # Operation parameters
+                pars = cmd.op.p
+                for reg in labels:
+                    # All the possible preparations should go in this loop
+                    if isinstance(cmd.op, Bosonic):
+                        weights, means, covs = [pars[i].tolist() for i in range(3)]
+
+                    elif isinstance(cmd.op, Catstate):
+                        weights, means, covs = self.prepare_cat(*pars)
+
+                    elif isinstance(cmd.op, GKP):
+                        weights, means, covs = self.prepare_gkp(*pars)
+
+                    elif isinstance(cmd.op, Fock):
+                        weights, means, covs = self.prepare_fock(*pars)
+
+                    elif type(cmd.op) in (Ket, DensityMatrix):
```

Comment:
Recommend changing all checks of this type to use `isinstance` :) Note that `isinstance` supports the second argument being a _sequence_ of allowed classes!

Reply:
```suggestion
                    elif isinstance(cmd.op, (Ket, DensityMatrix)):
```